src/sync/remote_mongo_result_types.hpp

```diff
@@ -0,0 +1,101 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#ifndef REMOTE_MONGO_RESULT_TYPES_HPP
+#define REMOTE_MONGO_RESULT_TYPES_HPP
+
+#include <json.hpp>
+
+namespace realm {
+namespace app {
+
+// MARK: - Coding keys
+
+const static std::string inserted_id_key = "insertedId";
+const static std::string inserted_ids_key = "insertedIds";
+const static std::string deleted_count_key = "deletedCount";
+const static std::string matched_count_key = "matchedCount";
+const static std::string modified_count_key = "modifiedCount";
+const static std::string upserted_id_key = "upsertedId";
+
+// TODO: Place using statement here for now until types exists
+using BSONValue = std::string;
+
+// MARK: - Result types
+
+/// Options to use when executing a `find` command on a `RemoteMongoCollection`.
+struct RemoteFindOptions {
+    /// The maximum number of documents to return.
+    util::Optional<long int> limit;
+
+    /// Limits the fields to return for all matching documents.
+    util::Optional<std::string> projection_json;
+
+    /// The order in which to return matching documents.
+    util::Optional<std::string> sort_json;
+};
+
+/// The result of an `insert_one` command on a `RemoteMongoCollection`.
+struct RemoteInsertOneResult {
+    /// The identifier that was inserted. If the document doesn't have an identifier, this value
+    /// will be generated by the Stitch server and added to the document before insertion.
+    BSONValue inserted_id;
+};
+
+/// The result of an `insert_many` command on a `RemoteMongoCollection`.
+struct RemoteInsertManyResult {
+    /// Map of the index of the inserted document to the id of the inserted document.
+    std::map<long int, BSONValue> inserted_ids;
+};
+
+/// The result of a `delete` command on a `RemoteMongoCollection`.
+struct RemoteDeleteResult {
+    /// The number of documents that were deleted.
+    int deleted_count;
+};
+
+/// The result of an `update_one` or `update_many` operation a `RemoteMongoCollection`.
+struct RemoteUpdateResult {
+    /// The number of documents that matched the filter.
+    int matched_count;
+    /// The number of documents modified.
+    int modified_count;
+    /// The identifier of the inserted document if an upsert took place.
+    BSONValue upserted_id;
+};
+
+/// Options to use when executing a `find_one_and_update`, `find_one_and_replace`,
+/// or `find_one_and_delete` command on a `remote_mongo_collection`.
+struct RemoteFindOneAndModifyOptions {
+    /// Limits the fields to return for all matching documents.
+    util::Optional<std::string> projection_json;
+    /// The order in which to return matching documents.
+    util::Optional<std::string> sort_json;
+    /// Whether or not to perform an upsert, default is false
+    /// (only available for find_one_and_replace and find_one_and_update)
+    util::Optional<bool> upsert;
```

Comment:
An optional bool gives three states; null, true, false. But I'm not sure we need that, if all you want to do is have a default in place, you can use `bool upsert = false;` Is that sufficient for how this will be used?

Reply:
I've removed this struct so we don't need to specify false. It has been moved to be a param on the method itself instead